skimage/filters/rank/core_cy.pyx

```diff
@@ -271,3 +286,293 @@ cdef void _core(void kernel(dtype_t_out*, Py_ssize_t, Py_ssize_t*, double,
+                        se_s_p[num_se_s] = p - centre_p
+                        num_se_s += 1
+
+        for p in range(planes):
+            for i in range(n_bins):
+                histo[i] = 0
+            pop = 0
+
+            for r in range(srows):
+                for c in range(scols):
+                    for j in range(splanes):
+                        rr = r - centre_r
+                        cc = c - centre_c
+                        pp = j - centre_p + p
+
+                        # change this to accept 3D
+                        if selem[j, r, c]:
+                            if is_in_mask_3D(rows, cols, planes, rr, cc, pp,
+                                             mask_data):
+                                histogram_increment(histo, &pop,
+                                                    image[pp, rr, cc])
+
+            r = 0
+            c = 0
+            kernel(&out[p, r, c, 0], odepth, histo, pop, image[p, r, c],
+                   n_bins, mid_bin, p0, p1, s0, s1)
+
+        # main loop
+
+            r = 0
+            for even_row in range(0, rows, 2):
+
+                # ---> west to east
+                for c in range(1, cols):
+                    for j in range(num_se_e):
+                        rr = r + se_e_r[j]
+                        cc = c + se_e_c[j]
+                        pp = p + se_e_p[j]
+                        if is_in_mask_3D(rows, cols, planes, rr, cc, pp,
+                                         mask_data):
+                            histogram_increment(histo, &pop, image[pp, rr, cc])
+
+                    for j in range(num_se_w):
+                        rr = r + se_w_r[j]
+                        cc = c + se_w_c[j] - 1
+                        pp = p + se_w_p[j]
+                        if is_in_mask_3D(rows, cols, planes, rr, cc, pp,
+                                         mask_data):
+                            histogram_decrement(histo, &pop, image[pp, rr, cc])
+
+                    kernel(&out[p, r, c, 0], odepth, histo, pop,
+                           image[p, r, c], n_bins, mid_bin, p0, p1, s0, s1)
+
+                r += 1  # pass to the next row
+                if r >= rows:
+                    break
+
+                # ---> north to south
+                for j in range(num_se_s):
+                    rr = r + se_s_r[j]
+                    cc = c + se_s_c[j]
+                    pp = p + se_s_p[j]
+                    if is_in_mask_3D(rows, cols, planes, rr, cc, pp,
+                                     mask_data):
+                        histogram_increment(histo, &pop, image[pp, rr, cc])
+
+                for j in range(num_se_n):
+                    rr = r + se_n_r[j] - 1
+                    cc = c + se_n_c[j]
+                    pp = p + se_n_p[j]
+                    if is_in_mask_3D(rows, cols, planes, rr, cc, pp,
+                                     mask_data):
+                        histogram_decrement(histo, &pop, image[pp, rr, cc])
```

Comment:
This pattern is repeated lots of times, where only `num_se_*` and `se_*_{p,r,c}` change, and the rest of the processing is the same. So, it could be a function! `_update_histogram(num_se_s, num_se_n, se_s, se_n, ...)`... And, if you store all the data in a bigger array where e/w/n/s are encoded as indices along an array axis, it can be even simpler, `_update_histogram(num_se, se, axis_inc=1, axis_dec=2)` (for example). In fact, since you are always decrementing/incrementing along the same direction, if you store the data in the position (e/n/w/s), you can always assume axis_inc=axis_dec + 2.

Does this all make sense? I think it would make the inner loop a lot easier to read...

Reply:
Sure, I can look into this. I guess I should also change the 2D version?

Reply:
Let's do the 3D version only for now, to keep the code changes as small as possible.

Reply:
The nice thing about all this, btw, is that I finally understand the 2D version, from having to review your PR. I never took the time to look at the 2D code very closely. =D